metric/registry/registry.go

```diff
@@ -25,52 +25,88 @@ import (
 
 // MeterProvider is a standard MeterProvider for wrapping `MeterImpl`
 type MeterProvider struct {
-	impl metric.MeterImpl
+	lock   sync.Mutex
+	impl   metric.MeterProvider
+	meters map[uniqueMeterKey]metric.Meter
 }
 
-var _ metric.MeterProvider = (*MeterProvider)(nil)
-
-// uniqueInstrumentMeterImpl implements the metric.MeterImpl interface, adding
-// uniqueness checking for instrument descriptors.  Use NewUniqueInstrumentMeter
-// to wrap an implementation with uniqueness checking.
-type uniqueInstrumentMeterImpl struct {
-	lock  sync.Mutex
-	impl  metric.MeterImpl
-	state map[key]metric.InstrumentImpl
+type uniqueMeterKey struct {
+	name      string
+	version   string
+	schemaURL string
 }
 
-var _ metric.MeterImpl = (*uniqueInstrumentMeterImpl)(nil)
-
-type key struct {
-	instrumentName         string
-	instrumentationName    string
-	InstrumentationVersion string
+func keyOf(name string, opts ...metric.MeterOption) uniqueMeterKey {
+	cfg := metric.NewMeterConfig(opts...)
+	return uniqueMeterKey{
+		name:      name,
+		version:   cfg.InstrumentationVersion(),
+		schemaURL: cfg.SchemaURL(),
+	}
 }
 
-// NewMeterProvider returns a new provider that implements instrument
+var _ metric.MeterProvider = (*MeterProvider)(nil)
+
+// NewMeterProvider returns a new provider that implements meter
 // name-uniqueness checking.
-func NewMeterProvider(impl metric.MeterImpl) *MeterProvider {
+func NewMeterProvider(impl metric.MeterProvider) *MeterProvider {
 	return &MeterProvider{
-		impl: NewUniqueInstrumentMeterImpl(impl),
+		impl:   impl,
+		meters: map[uniqueMeterKey]metric.Meter{},
 	}
 }
 
 // Meter implements MeterProvider.
 func (p *MeterProvider) Meter(instrumentationName string, opts ...metric.MeterOption) metric.Meter {
-	return metric.WrapMeterImpl(p.impl, instrumentationName, opts...)
+	k := keyOf(instrumentationName, opts...)
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	m, ok := p.meters[k]
+	if !ok {
+		m = metric.WrapMeterImpl(NewMeterImpl(
+			p.impl.Meter(instrumentationName, opts...).MeterImpl(),
+		))
+		p.meters[k] = m
+	}
+	return m
+
 }
 
+// List provides a list of MeterImpl objects created through this
+// provider.
+func (p *MeterProvider) List() []metric.MeterImpl {
+	p.lock.Lock()
+	defer p.lock.Unlock()
```

Comment:
Since this code moved into the controller, now it's using the controller's lock. This required a slight change in `Shutdown()` to release the lock before the final collection.